Shaders/ssrefr_pass/ssrefr_pass.frag.glsl

```diff
@@ -0,0 +1,126 @@
+//refraction from modified reflection by Yvain.
+#version 450
+
+#include "compiled.inc"
+#include "std/math.glsl"
+#include "std/gbuffer.glsl"
+
+in vec2 texCoord;
+out vec4 fragColor;
+
+uniform sampler2D tex;
+uniform sampler2D tex1;
+uniform sampler2D gbufferD;
+uniform sampler2D gbuffer0;
+uniform sampler2D gbufferD1;
+uniform sampler2D gbuffer_refraction; //ior\opacity
+uniform mat4 P;
+uniform mat3 V3;
+uniform vec2 cameraProj;
+uniform vec3 eye;
+uniform vec3 eyeLook;
+
+#ifdef _CPostprocess
+uniform vec3 PPComp9;
+uniform vec3 PPComp10;
+#endif
+
+in vec3 viewRay;
+vec3 hitCoord;
+float depth;
+vec3 viewPos;
+
+const int numBinarySearchSteps = 7;
+const int maxSteps = 18;
+
+vec2 getProjectedCoord(const vec3 hit) {
+	vec4 projectedCoord = P * vec4(hit, 1.0);
+	projectedCoord.xy /= projectedCoord.w;
+	projectedCoord.xy = projectedCoord.xy * 0.5 + 0.5;
+	#ifdef _InvY
+	projectedCoord.y = 1.0 - projectedCoord.y;
+	#endif
+	return projectedCoord.xy;
+}
+
+float getDeltaDepth(const vec3 hit) {
+	float depth2 = textureLod(gbufferD1, getProjectedCoord(hit), 0.0).r * 2.0 - 1.0;
+	return depth2;
+}
+
+vec4 binarySearch(vec3 dir) {
+	float d;
+	for (int i = 0; i < numBinarySearchSteps; i++) {
+		dir *= ss_refractionMinRayStep;
+		hitCoord += dir;
+		d = getDeltaDepth(hitCoord);
+		if (d < depth)
+			hitCoord -= dir;
+	}
+	return vec4(getProjectedCoord(hitCoord), 0.0, 1.0);
+}
+
+vec4 rayCast(vec3 dir) {
+	#ifdef _CPostprocess
+	dir *= PPComp9.x;
+	#else
+	dir *= ss_refractionRayStep;
+	#endif
+	for (int i = 0; i < maxSteps; i++) {
+		hitCoord += dir;
+		float d = getDeltaDepth(hitCoord);
+		if(d > depth)
+			return binarySearch(dir);
+	}
+	return vec4(texCoord, 0.0, 1.0);
+}
+
+void main() {
+	vec4 g0 = textureLod(gbuffer0, texCoord, 0.0);
+	float roughness = g0.z;
+	vec4 gr = textureLod(gbuffer_refraction, texCoord, 0.0);
+	float ior = unpackFloat2(gr.r).x;
+	float opac = unpackFloat2(gr.r).y;
+
+	depth = textureLod(gbufferD, texCoord, 0.0).r * 2.0 - 1.0;
+
+	if (depth == 1.0) { fragColor = textureLod(tex, texCoord, 0.0); return; }
+	else if (ior == 1.0 || opac == 1.0 || roughness == 1.0) { 
+		discard; return;
```

Comment:
I don't think you need the additional `return` here

Reply:
yep